README.md
sk_fancybox
===========

This extension integrates the jQuery plugin FancyBox into TYPO3. 


ChangeLog
2011-03-20 Samuel Koch  <sam@koch.is>

    * Added note to the manual concerning inclusion through "includeLib".

2011-01-17 Samuel Koch  <sam@koch.is>

    * Fixed the manual.

2011-01-14 Samuel Koch  <sam@koch.is>

    * Additional title position added.

2011-01-13 Samuel Koch  <sam@koch.is>

    * Added the manual.
    * JS and CSS files are now only included if static template is included as well.

2010-12-29 Samuel Koch  <sam@koch.is>

    * Some small enhancements.

2010-12-28 Samuel Koch  <sam@koch.is>

    * Some enhancements.

2010-12-28 Samuel Koch  <sam@koch.is>

    * Initial release with FancyBox v. 1.3.4.

class.tx_skfancybox.php
<?php
    /***************************************************************
    *  Copyright notice
    *
    *  (c) 2010 Samuel Koch <sam@koch.is>
    *  All rights reserved
    *
    *  This script is part of the TYPO3 project. The TYPO3 project is
    *  free software; you can redistribute it and/or modify
    *  it under the terms of the GNU General Public License as published by
    *  the Free Software Foundation; either version 2 of the License, or
    *  (at your option) any later version.
    *
    *  The GNU General Public License can be found at
    *  http://www.gnu.org/copyleft/gpl.html.
    *
    *  This script is distributed in the hope that it will be useful,
    *  but WITHOUT ANY WARRANTY; without even the implied warranty of
    *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    *  GNU General Public License for more details.
    *
    *  This copyright notice MUST APPEAR in all copies of the script!
    ***************************************************************/
    /**
    * [CLASS/FUNCTION INDEX of SCRIPT]
    *
    * Hint: use extdeveval to insert/update function index above.
    */

    //require_once(PATH_tslib.'class.tslib_pibase.php');


    /**
    * Plugin 'FancyBox' for the 'sk_fancybox' extension.
    *
    * @author    Samuel Koch <sam@koch.is>
    * @package    TYPO3
    * @subpackage    tx_skfancybox
    */
    class tx_skfancybox extends tslib_pibase {
        var $prefixId      = 'tx_skfancybox';        // Same as class name
        var $scriptRelPath = 'class.tx_skfancybox.php';    // Path to this script relative to the extension dir.
        var $extKey        = 'sk_fancybox';    // The extension key.
        var $pi_checkCHash = true;
        var $extPath;

        /**
        * The main method of the PlugIn
        *
        * @param    string        $content: The PlugIn content
        * @param    array        $conf: The PlugIn configuration
        * @return    The content that is displayed on the website
        */
        function main($content,$conf) {
            $this->conf = $conf;

            //only execute this function if static template is included
            if(!$this->conf['static_template_included']) {
                return false;
            }

            $this->pi_loadLL();
            $this->extPath = t3lib_extMgm::siteRelPath($this->extKey);

            //get extconf values
            $extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['sk_fancybox']);

            //check whether to include jQuery library
            if (!$extConf['dontIncludeJquery']) {
                //check whether extension t3jquery is loaded
                if (t3lib_extMgm::isLoaded('t3jquery')) {
                    require_once(t3lib_extMgm::extPath('t3jquery').'class.tx_t3jquery.php');
                }
                //if t3jquery is loaded and the custom library has been created
                if (T3JQUERY === true) {
                    tx_t3jquery::addJqJS();
                }
                //otherwise include jQuery file
                else {
                    $addJsFiles[] = $this->extPath.'fancybox/jquery-1.4.4.min.js';
                }
            }

            //check whether to include jQuery easing
            if ($extConf['includeJqueryEasing']) {
                $addJsFiles[] = $this->extPath.'fancybox/jquery.easing-1.3.pack.js';
            }
            //check whether to include jQuery mousewheel
            if ($extConf['includeJqueryMousewheel']) {
                $addJsFiles[] = $this->extPath.'fancybox/jquery.mousewheel-3.0.4.pack.js';
            }

            //check whether to include the FancyBox library
            if (!$extConf['dontIncludeFancyboxJs']) {
                $addJsFiles[] = $this->extPath.'fancybox/jquery.fancybox-1.3.4.pack.js';
            }

            //check whether to include the FancyBox stylesheet
            if (strlen($this->conf['stylesheetPath']) > 0) {
                $addCssFiles[] = $this->conf['stylesheetPath'];
            }

            //check TYPO3 version to see whether pagerenderer is available
            if(class_exists('t3lib_utility_VersionNumber')) {
                $typo3Version = t3lib_utility_VersionNumber::convertVersionNumberToInteger(TYPO3_version);
            }
            else {
                $typo3Version = t3lib_div::int_from_ver(TYPO3_version);
            }
            //check TYPO3 version to see whether pagerenderer is available
            if ($typo3Version >= 4003000) {
                //get pagerenderer
                $pagerender = $GLOBALS['TSFE']->getPageRenderer();

                //add JS and CSS files
                foreach($addJsFiles as $jsFile) {
                    $pagerender->addJsFile($jsFile);
                }
                foreach($addCssFiles as $CssFile) {
                    $pagerender->addCssFile($CssFile);
                }
            }
            //if pagerenderer is not available add JS and CSS files the old way
            else {
                foreach($addJsFiles as $jsFile) {
                    $headerParts .= '<script type="text/javascript" src="'.$jsFile.'"></script>';
                }
                foreach($addCssFiles as $CssFile) {
                    $headerParts .= '<link rel="stylesheet" href="'.$CssFile.'" type="text/css" />';
                }
                $GLOBALS['TSFE']->additionalHeaderData['EXT:' . $this->extKey] = $headerParts;
            }


            //process display of the title
            if($this->conf['titlePosition'] == "over") {
                $titleFormat = '
                function(title, currentArray, currentIndex, currentOpts) {
                if(title.length > 0 || currentArray.length > 1) {
                var preCaption = \'\';
                if(currentArray.length > 1) {
                preCaption = \''.$this->pi_getLL('image').' \' + (currentIndex + 1) + \' '.$this->pi_getLL('of').' \' + currentArray.length + (title.length ? \' : &nbsp; \' : \'\');
                }
                return \'<span id="fancybox-title-over">\' + preCaption + title + \'</span>\';
                }
                return false;
                }';
                $titleFormat = str_replace(array("\r", "\n"), '', $titleFormat);
            }
            else {
                $titleFormat = '"formatTitle"';
            }

            //set FancyBox options
            $GLOBALS['TSFE']->inlineJS['sk_fancybox'] = '
            jQuery(document).ready(function() {

            jQuery(".fancybox").fancybox({
            "padding": '.$this->conf['padding'].',
            "margin": '.$this->conf['margin'].',
            "opacity": '.($this->conf['opacity'] ? 'true' : 'false').',
            "modal": '.($this->conf['modal'] ? 'true' : 'false').',
            "cyclic": '.($this->conf['cyclic'] ? 'true' : 'false').',
            "scrolling": "'.$this->conf['scrolling'].'",
            "autoScale": '.($this->conf['autoScale'] ? 'true' : 'false').',
            "autoDimensions": '.($this->conf['autoDimensions'] ? 'true' : 'false').',
            "centerOnScroll": '.($this->conf['centerOnScroll'] ? 'true' : 'false').',
            "hideOnOverlayClick": '.($this->conf['hideOnOverlayClick'] ? 'true' : 'false').',
            "hideOnContentClick": '.($this->conf['hideOnContentClick'] ? 'true' : 'false').',
            "overlayShow": '.($this->conf['overlayShow'] ? 'true' : 'false').',
            "overlayOpacity": '.$this->conf['overlayOpacity'].',
            "overlayColor": "'.$this->conf['overlayColor'].'",
            "titleShow": '.($this->conf['titleShow'] ? 'true' : 'false').',
            '.($this->conf['titlePosition'] != 'outside-bar' ? '"titlePosition": "'.$this->conf['titlePosition'].'",' : '').'
            "titleFormat": '.$this->conf['titleFormat'].',
            "transitionIn": "'.$this->conf['transitionIn'].'",
            "transitionOut": "'.$this->conf['transitionOut'].'",
            "speedIn": '.$this->conf['speedIn'].',
            "speedOut": '.$this->conf['speedOut'].',
            "changeSpeed": '.$this->conf['changeSpeed'].',
            "changeFade": "'.$this->conf['changeFade'].'",
            "easingIn": "'.$this->conf['easingIn'].'",
            "easingOut": "'.$this->conf['easingOut'].'",
            "showCloseButton": '.($this->conf['showCloseButton'] ? 'true' : 'false').',
            "showNavArrows": '.($this->conf['showNavArrows'] ? 'true' : 'false').',
            "enableEscapeButton": '.($this->conf['enableEscapeButton'] ? 'true' : 'false').',
            "titleFormat": '.$titleFormat.'
            });

            });';
        }


        /**
        * Function that sets the register var "IMAGE_NUM_CURRENT" to the the current image number.
        *
        * BEWARE: Since tt_news 3.0 this won't work until Rupert updates hooks for marker-processing
        *
        * @param    array        $paramArray: $markerArray and $config of the current news item in an array
        * @param    [type]        $conf: ...
        * @return    array        the processed markerArray
        */
        function ttnewsImageMarkerFunc($paramArray,$conf) {
            $markerArray = $paramArray[0];
            $lConf = $paramArray[1];
            $pObj = &$conf['parentObj'];
            $row = $pObj->local_cObj->data;



            $imageNum = isset($lConf['imageCount']) ? $lConf['imageCount']:1;
            $theImgCode = '';
            $imgs = t3lib_div::trimExplode(',', $row['image'], 1);
            $imgsCaptions = explode(chr(10), $row['imagecaption']);
            $imgsAltTexts = explode(chr(10), $row['imagealttext']);
            $imgsTitleTexts = explode(chr(10), $row['imagetitletext']);

            reset($imgs);

            $cc = 0;

            // BEN: We need to mark these items prior to arrayshifting
            if (count($imgs) == 1 &&
                $pObj->config['firstImageIsPreview'] &&
                $pObj->config['code'] == 'SINGLE' &&
                !$pObj->config['forceFirstImageIsPreview'])
            {
                $markedAsSpecial = 1;
            }
            // END.

            // remove first img from the image array in single view if the TSvar firstImageIsPreview is set
            if ((    (count($imgs) > 1 && $pObj->config['firstImageIsPreview'])
                ||
                (count($imgs) >= 1 && $pObj->config['forceFirstImageIsPreview'])
                ) && $pObj->config['code'] == 'SINGLE') {
                array_shift($imgs);
                array_shift($imgsCaptions);
                array_shift($imgsAltTexts);
                array_shift($imgsTitleTexts);
            }
            // get img array parts for single view pages
            if ($this->piVars[$pObj->config['singleViewPointerName']]) {
                $spage = $this->piVars[$pObj->config['singleViewPointerName']];
                $astart = $imageNum*$spage;
                $imgs = array_slice($imgs,$astart,$imageNum);
                $imgsCaptions = array_slice($imgsCaptions,$astart,$imageNum);
                $imgsAltTexts = array_slice($imgsAltTexts,$astart,$imageNum);
                $imgsTitleTexts = array_slice($imgsTitleTexts,$astart,$imageNum);
            }

            while (list(, $val) = each($imgs)) {
                if ($cc == $imageNum) break;
                if ($val) {

                    $lConf['image.']['altText'] = $imgsAltTexts[$cc];
                    $lConf['image.']['titleText'] = $imgsTitleTexts[$cc];
                    $lConf['image.']['file'] = 'uploads/pics/' . $val;
                    // BEN: We check count of images >(=) 0 here because the array got shifted before!!! (See above)
                    if ((    (count($imgs) > 0 && $pObj->config['firstImageIsPreview'])
                        ||
                        (count($imgs) >= 0 && $pObj->config['forceFirstImageIsPreview'])
                        ) && $pObj->config['code'] == 'SINGLE') {
                        // BEN: Additionally we need to check our special case
                        if (count($imgs) == 1 && $markedAsSpecial) {
                            $GLOBALS['TSFE']->register['IMAGE_NUM_CURRENT'] = $cc;
                        } else {
                            $GLOBALS['TSFE']->register['IMAGE_NUM_CURRENT'] = $cc+1;
                        }
                    } else {
                        $GLOBALS['TSFE']->register['IMAGE_NUM_CURRENT'] = $cc;
                    }
                    // END.
                }

                $theImgCode .= $pObj->local_cObj->IMAGE($lConf['image.']) . $pObj->local_cObj->stdWrap($imgsCaptions[$cc], $lConf['caption_stdWrap.']);
                $cc++;
            }
            $markerArray['###NEWS_IMAGE###'] = '';
            if ($cc) {
                $markerArray['###NEWS_IMAGE###'] = $pObj->local_cObj->wrap(trim($theImgCode), $lConf['imageWrapIfAny']);
            } else {
                $markerArray['###NEWS_IMAGE###'] = $pObj->local_cObj->stdWrap($markerArray['###NEWS_IMAGE###'],$lConf['image.']['noImage_stdWrap.']);
            }
            return $markerArray;
        }

    }



    if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/sk_fancybox/class.tx_skfancybox.php'])    {
        include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/sk_fancybox/class.tx_skfancybox.php']);
    }

?>

ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "sk_fancybox".
 *
 * Auto generated 18-03-2013 12:52
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'FancyBox',
	'description' => 'This extension integrates the jQuery plugin FancyBox into TYPO3. FancyBox is a fancy lightbox alternative. It replaces the "Enlarge on Click" functionality with the FancyBox (supports tt_news). See http://www.fancybox.net/ for demos.',
	'category' => 'fe',
	'shy' => 0,
	'version' => '1.0.4',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'stable',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Samuel Koch',
	'author_email' => 'sam@koch.is',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.4.0-6.0.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:37:{s:9:"ChangeLog";s:4:"8fc9";s:23:"class.tx_skfancybox.php";s:4:"e881";s:21:"ext_conf_template.txt";s:4:"7b40";s:12:"ext_icon.gif";s:4:"363e";s:17:"ext_localconf.php";s:4:"3350";s:14:"ext_tables.php";s:4:"0cdd";s:13:"locallang.xml";s:4:"9d1e";s:10:"README.txt";s:4:"128f";s:14:"doc/manual.sxw";s:4:"1ac9";s:18:"fancybox/blank.gif";s:4:"3254";s:24:"fancybox/fancy_close.png";s:4:"6e28";s:26:"fancybox/fancy_loading.png";s:4:"b1d5";s:27:"fancybox/fancy_nav_left.png";s:4:"3f3e";s:28:"fancybox/fancy_nav_right.png";s:4:"216e";s:27:"fancybox/fancy_shadow_e.png";s:4:"fd4f";s:27:"fancybox/fancy_shadow_n.png";s:4:"18cd";s:28:"fancybox/fancy_shadow_ne.png";s:4:"63ad";s:28:"fancybox/fancy_shadow_nw.png";s:4:"c820";s:27:"fancybox/fancy_shadow_s.png";s:4:"9b9e";s:28:"fancybox/fancy_shadow_se.png";s:4:"a8af";s:28:"fancybox/fancy_shadow_sw.png";s:4:"f81c";s:27:"fancybox/fancy_shadow_w.png";s:4:"59b0";s:29:"fancybox/fancy_title_left.png";s:4:"1582";s:29:"fancybox/fancy_title_main.png";s:4:"38da";s:29:"fancybox/fancy_title_over.png";s:4:"b886";s:30:"fancybox/fancy_title_right.png";s:4:"6cbe";s:23:"fancybox/fancybox-x.png";s:4:"1686";s:23:"fancybox/fancybox-y.png";s:4:"36a5";s:21:"fancybox/fancybox.png";s:4:"11e5";s:28:"fancybox/jquery-1.4.4.min.js";s:4:"73a9";s:34:"fancybox/jquery.easing-1.3.pack.js";s:4:"def2";s:34:"fancybox/jquery.fancybox-1.3.4.css";s:4:"c2d4";s:33:"fancybox/jquery.fancybox-1.3.4.js";s:4:"e7fc";s:38:"fancybox/jquery.fancybox-1.3.4.pack.js";s:4:"8bc3";s:40:"fancybox/jquery.mousewheel-3.0.4.pack.js";s:4:"3b0a";s:20:"static/constants.txt";s:4:"7c17";s:16:"static/setup.txt";s:4:"5d21";}',
	'suggests' => array(
	),
);

?>